### PR TITLE
Add iomux rt driver

### PR DESCRIPTION
**Pull request by hakehuang**
add iomux_rt driver for io mux settings, which will be used for sai

### CODE REVIEW

**Review by gmarull — CHANGES_REQUESTED** [review]
No new pinmux drivers should be added, please use pinctrl API. Ref. https://github.com/zephyrproject-rtos/zephyr/issues/39740

**Comment by hakehuang** [comment]
@gmarull , according to the plan pinctl will implement on 3.0.1 release for imx rt series, can we have this driver in first?

**Comment by gmarull** [comment]
> @gmarull , according to the plan pinctl will implement on 3.0.1 release for imx rt series, can we have this driver in first?

The agreement was to not allow any new pinmux based drivers. If you consider that you need an exception, please bring this to the API or TSC meetings.

**Comment by dleach02** [comment]
@gmarull, while I understand the desire to avoid adding additional drivers, we (NXP) have it in our plans to do the transitioning of all of our in-tree platforms after the 3.0 release to complete by 3.1. For various reasons, we need to get this particular PR in for some follow up items we are attempting to get in before 3.0.

File: boards/arm/mimxrt1010_evk/mimxrt1010_evk_defconfig [hunk 1]
```diff
@@ -12,3 +12,5 @@ CONFIG_SERIAL=y
 CONFIG_GPIO=y
 CONFIG_ARM_MPU=y
 CONFIG_HW_STACK_PROTECTION=y
+CONFIG_PINMUX=y
+CONFIG_PINMUX_MCUX_RT=y
```
Comment (galak) [inline boards/arm/mimxrt1010_evk/mimxrt1010_evk_defconfig:16]: Can you update Kconfig here to just set `CONFIG_PINMUX=y` and drop `CONFIG_PINMUX_MCUX_RT=y` from all the defconfigs.

Add something in the SoC like we do for CONFIG_GPIO.

**Comment by galak** [comment]
Dev-review (Dec 16, 2021):
* What needs this driver - want to understand the need here.  Is this just to enable SAI driver as it already calls `pinmux_pin_set`?
* Request to allow this in from @dleach02 and NXP will work on fixing up all NXP related drivers to new pinmux in early 2022.

File: boards/arm/mimxrt1010_evk/mimxrt1010_evk_defconfig (hunk 1 above)
Comment (hakehuang) [reply boards/arm/mimxrt1010_evk/mimxrt1010_evk_defconfig:16]: update.

**Comment by gmarull** [comment]
One of the questions I have: why not use a board `pinmux.c` file to configure pins while a pinctrl driver is not in place? I see boards like `mimxrt1015_evk` that do that. What has changed to now need a pinmux driver?

**Comment by hakehuang** [comment]
> One of the questions I have: why not use a board `pinmux.c` file to configure pins while a pinctrl driver is not in place? I see boards like `mimxrt1015_evk` that do that. What has changed to now need a pinmux driver?

the iomux here is mclk which is not a fixed one, it needs programable according to status.

**Comment by gmarull** [comment]
> > One of the questions I have: why not use a board `pinmux.c` file to configure pins while a pinctrl driver is not in place? I see boards like `mimxrt1015_evk` that do that. What has changed to now need a pinmux driver?
> 
> the iomux here is mclk which is not a fixed one, it needs programable according to status.

Can you please provide usecases for this proposal?

**Comment by hakehuang** [comment]
> Can you please provide usecases for this proposal?

in our i2s sai IP , the mclk can be as below, so if the codec can provide a better clock source , then we can use this clock as mclk.

image

**Comment by gmarull** [comment]
> > Can you please provide usecases for this proposal?
> 
> in our i2s sai IP , the mclk can be as below, so if the codec can provide a better clock source , then we can use this clock as mclk.
> 
> image

This doesn't answer the original question: why this can't be handled in a board pinmux.c file (or in the driver directly)? Can you please provide code example? maybe a PR using this API?

**Comment by hakehuang** [comment]
> This doesn't answer the original question: why this can't be handled in a board pinmux.c file (or in the driver directly)? Can you please provide code example? maybe a PR using this API?

the PR is merged in i2s_mcux_sai.c line 329
```
static void enable_mclk_direction(const struct device *dev, bool dir)
{
	const struct i2s_mcux_config *dev_cfg = dev->config;
	const struct device *iomuxgpr_dev = dev_cfg->pinmux_dev;
	uint32_t offset = dev_cfg->mclk_pin_offset;
	uint32_t mask = dev_cfg->mclk_pin_mask;
	uint32_t value = 0;

	if (iomuxgpr_dev == NULL) {
		LOG_ERR("fail to find pinmux controller\n");
		return;
	}

	pinmux_pin_get(iomuxgpr_dev, offset, &value);
	if (dir) {
		/* enable MCLK output */
		value |= mask;
	} else {
		/* disable MCLK output */
		value &= ~mask;
	}
	pinmux_pin_set(iomuxgpr_dev, offset, value);
}

```

**Comment by gmarull** [comment]
> > This doesn't answer the original question: why this can't be handled in a board pinmux.c file (or in the driver directly)? Can you please provide code example? maybe a PR using this API?
> 
> the PR is merged in i2s_mcux_sai.c line 329
> 
> ```
> static void enable_mclk_direction(const struct device *dev, bool dir)
> {
> 	const struct i2s_mcux_config *dev_cfg = dev->config;
> 	const struct device *iomuxgpr_dev = dev_cfg->pinmux_dev;
> 	uint32_t offset = dev_cfg->mclk_pin_offset;
> 	uint32_t mask = dev_cfg->mclk_pin_mask;
> 	uint32_t value = 0;
> 
> 	if (iomuxgpr_dev == NULL) {
> 		LOG_ERR("fail to find pinmux controller\n");
> 		return;
> 	}
> 
> 	pinmux_pin_get(iomuxgpr_dev, offset, &value);
> 	if (dir) {
> 		/* enable MCLK output */
> 		value |= mask;
> 	} else {
> 		/* disable MCLK output */
> 		value &= ~mask;
> 	}
> 	pinmux_pin_set(iomuxgpr_dev, offset, value);
> }
> ```

and how is that code supposed to work without this PR? does `main` work now?

**Comment by gmarull** [comment]
I think that the i2s driver can be modified like this:

```c
uint32_t *gpr = (uint32_t *)DT_REG_ADDR(DT_NODELABEL(iomuxcgpr)) + offset; 
if (dir) {
    *gpr |= mask;
} else {
    *gpr &= ~mask;
}
```

and you're done until a proper pinctrl driver is in place.

**Comment by hakehuang** [comment]
> and you're done until a proper pinctrl driver is in place.

let me create PR accordingly. Thanks

**Comment by hakehuang** [comment]
@gmarull move the recommend changed in #41159 41159

**Pull request closed without merging**